### PR TITLE
Hard Limit the number of hop stream goroutines

### PR DESCRIPTION
**Pull request by vyzo**
This adds a hard limit to the number of hop goroutines, so that relays don't get overloaded.

Note that the live hop tracking has been removed for two reasons:
- it was not by any way exported to the outside world, just taking memory
- there is lock contention

Note 2: DO NOT MERGE AS IS; I will rebase/squash before merging to clean up history.

### CODE REVIEW

**Comment by vyzo** [comment]
An alternative to hard resetting is to add a new error code for overloaded relays.
But if we do this, we must stop lingering awaiting EOF in `handleError`; perhaps we should just reset after sending the error.

**Comment by vyzo** [comment]
Resetting the stream won't work, the error will not be propagated further.
But we _can_ reduce the EOF timeout to something much more reasonable than the current 1 minute (to say 5s).
Fortunately this is a variable in `go-libp2p-net`, so the consumer can set it.

**Comment by vyzo** [comment]
Implemented the `RELAY_OVERLOADED` error code, so that we don't do hard resets without notifying the other side. 
The relay daemon can set the `EOFTimeout` to something shorter in order to reduce the goroutine linger time.

**Comment by vyzo** [comment]
I reverted to resetting the stream when the hop limit is exceeded -- being nice has deleterious effects in the number of lingering goroutines.

File: relay.go [hunk 1]
```diff
@@ -47,9 +51,9 @@ type Relay struct {
 	relays map[peer.ID]struct{}
 	mx     sync.Mutex
 
-	liveHops map[peer.ID]map[peer.ID]int
-	lhCount  uint64
-	lhLk     sync.Mutex
+	// atomic counters
+	sCount  int32
```
Comment (Stebalien) [inline relay.go]: Nit: could we give these full names? (`streamCount`?)

File: relay.go [hunk 2]
```diff
@@ -29,6 +30,9 @@ var (
 	RelayAcceptTimeout   = 10 * time.Second
 	HopConnectTimeout    = 30 * time.Second
 	StopHandshakeTimeout = 1 * time.Minute
+
+	HopStreamBuffer = 4096
```
Comment (Stebalien) [inline relay.go]: ultra nit: `HopStreamBufferSize`

File: pb/relay.proto [hunk 3]
```diff
@@ -21,6 +21,7 @@ message CircuitRelay {
     STOP_DST_MULTIADDR_INVALID = 351;
     STOP_RELAY_REFUSED         = 390;
     MALFORMED_MESSAGE          = 400;
+    RELAY_OVERLOADED           = 500;
```
Comment (Stebalien) [inline pb/relay.proto]: Are we going to use this or should we just leave it at a reset?

File: relay.go [hunk 4]
```diff
@@ -29,6 +30,9 @@ var (
...
+	HopStreamLimit  = 1 << 18 // 256K hops for 512K goroutines
```
Comment (Stebalien) [inline relay.go]: Have we checked this against our current numbers? This seems kind of low, actually. For 20k peers, this'll give us less than 10 streams per peer.

File: relay.go (hunk 1 above)
Comment (vyzo) [reply relay.go]: Sure, will do.

File: relay.go (hunk 2 above)
Comment (vyzo) [reply relay.go]: ok!

File: pb/relay.proto (hunk 3 above)
Comment (vyzo) [reply pb/relay.proto]: I will drop it in the rebase/squash.

File: relay.go (hunk 4 above)
Comment (vyzo) [reply relay.go]: We can easily double it -- in fact the mplex relay where I am testing this is running with double the count (it overrides in the daemon init).

**Comment by vyzo** [comment]
> Sounds like something we need although I'm a bit worried we should be using per-peer limits instead. At the moment, ~500 peers could fully mesh-connect through the relay to kill it.

Per-peer limits are a little complex to implement, and need a lock (which I would like to avoid).
We have not observed fully connected meshes in the relays whatsoever; what has been observed is a long tail distribution where most peers have a small number of streams (they are connecting to the peers behind the relay) and then a bunch have a large number of streams (they are accepting connections).

File: relay.go (hunk 4 above)
Comment (vyzo) [reply relay.go]: I'll try with quadruple the count (hop limit at 1M) and evaluate memory usage with it.

Comment (vyzo) [reply relay.go]: Seems like we are tight on memory with 2M goroutines active.

Comment (vyzo) [reply relay.go]: I doubled the default, and actual daemons can set it higher if they have the resources.

File: relay.go (hunk 1 above)
Comment (vyzo) [reply relay.go]: done.

File: relay.go (hunk 2 above)
Comment (vyzo) [reply relay.go]: done

**Comment by Stebalien** [comment]
> Per-peer limits are a little complex to implement, and need a lock (which I would like to avoid).
We have not observed fully connected meshes in the relays whatsoever; what has been observed is a long tail distribution where most peers have a small number of streams (they are connecting to the peers behind the relay) and then a bunch have a large number of streams (they are accepting connections).

I'm primarily worried about someone attacking a relay this way but this certainly isn't the only way.

**Comment by vyzo** [comment]
rebased/squashed to just 2 commits and dropped the `RELAY_OVERLOADED` change in pb.

**Pull request merged**